File: julia/benchmark.jl
""" Wired.jl - Julia Benchmark Run File 
    (c) 2024 ryan@freestatelabs.com 
"""

include("wired.jl")
using BenchmarkTools, PrettyTables

N = (100, 1_000, 10_000) 
results = Matrix{Any}(undef, 5, 4)
header = ["Opt Level", "100x100", "1000x1000", "10000x10000"]
results[:,1] = ["Simple Loops", "Vectorization", "Parallel, 2 cores", "Parallel, 4 cores", "Parallel, 8 cores"]

for i in eachindex(N)
    print("Benchmarking with N = ")
    print(N[i]) 
    print("\n")
    nodes, sources = createproblem(N[i],N[i])
    results[1,i+1] = @belapsed solve($nodes, $sources)
    results[2,i+1] = @belapsed solve2($nodes, $sources)
    results[3,i+1] = @belapsed solve3($nodes, $sources, Nt=2)
    results[4,i+1] = @belapsed solve3($nodes, $sources, Nt=4)
    results[5,i+1] = @belapsed solve3($nodes, $sources, Nt=8)
end
println("Benchmarking complete.")
pretty_table(results, header=header)


# Display the answer to the problem using each function to ensure accuracy of 
#   the solution
nodes, sources = createproblem(100,100)
display(solve(nodes, sources))
display(solve2(nodes, sources))
display(solve3(nodes, sources))

File: python/benchmark.py
""" Wired.jl benchmark: Python implementation 
    (c) 2024 ryan@freestatelabs 

    Run this file from the command line with the following arguments: 
    $ python3 benchmark.py NUM_NODES NUM_SOURCES NUM_ITERATIONS OPT_LEVEL NUM_THREADS

    OPT_LEVEL definitions:
    0 : Python
    1 : NumPy loops 
    2 : NumPy, vectorized
    3 : Numba 
    4 : Numba, parallel 
"""
from wired import * 
import time, sys

def runtest(N=1000, M=1000, Ni=10, opt_level=2, Nt=1): 
    set_num_threads(Nt)
    times = np.zeros((Ni,1))

    # Create the problem
    solver = solvers[opt_level]
    if opt_level > 0: 
        nodes, sources = create_problem2(N, M)
    else:
        nodes, sources = create_problem(N, M)
    
    if opt_level > 2:
        print("Precompiling...")
        t0 = time.perf_counter()
        _ = solver(nodes[0:2,:], sources[0:2,:])    # Precompile and throw out result
        t1 = time.perf_counter()
        print("Precompilation complete. Precompilation time: {0:.3E} s".format(t1-t0))

    for i in range(0, Ni): 

        t0 = time.perf_counter()
        B = solver(nodes, sources) 
        t1 = time.perf_counter()
        times[i] = t1 - t0

    print("First 10 rows of B: ")
    if opt_level > 0: 
        print(B[:,0:10])
    else:
        print(B[0:10])
    avg_time = np.mean(times)
    avg_time_per_DOF = avg_time/(N*M)

    print("Running tests with opt level {0}.".format(opt_level))
    print(" Iterations:   {0}".format(Ni))
    print(" Nodes:        {0}".format(N))
    print(" Sources:      {0}".format(M))
    print(" Threads:      {0}".format(Nt))
    print(" Avg time:     {0:.3E} s".format(avg_time))
    print(" Avg time/DOF: {0:.3E} s".format(avg_time_per_DOF))
    print("Tests complete.")


# Run the program 
if __name__ == "__main__":
    a = [int(x) for x in sys.argv[1:]]
    runtest(a[0], a[1], a[2], a[3], a[4])